File: sdk/python/term_sdk/cost.py
"""
Cost tracking for LLM usage
"""

from dataclasses import dataclass, field
from datetime import datetime
from typing import Dict, List, Optional, Tuple

# Model pricing in USD per million tokens (input, output)
MODEL_PRICING: Dict[str, Tuple[float, float]] = {
    # OpenAI models
    "openai/gpt-4o": (2.50, 10.00),
    "openai/gpt-4o-mini": (0.15, 0.60),
    "openai/gpt-4-turbo": (10.00, 30.00),
    "openai/gpt-4": (30.00, 60.00),
    "openai/gpt-3.5-turbo": (0.50, 1.50),
    "openai/o1-preview": (15.00, 60.00),
    "openai/o1-mini": (3.00, 12.00),
    
    # Anthropic models
    "anthropic/claude-3.5-sonnet": (3.00, 15.00),
    "anthropic/claude-3-sonnet": (3.00, 15.00),
    "anthropic/claude-3-haiku": (0.25, 1.25),
    "anthropic/claude-3-opus": (15.00, 75.00),
    "anthropic/claude-2": (8.00, 24.00),
    
    # Meta models
    "meta-llama/llama-3.1-70b-instruct": (0.52, 0.75),
    "meta-llama/llama-3.1-8b-instruct": (0.055, 0.055),
    "meta-llama/llama-3-70b-instruct": (0.52, 0.75),
    
    # Mistral models
    "mistralai/mistral-large": (2.00, 6.00),
    "mistralai/mistral-medium": (2.70, 8.10),
    "mistralai/mistral-small": (0.20, 0.60),
    "mistralai/mixtral-8x7b-instruct": (0.24, 0.24),
    
    # Google models
    "google/gemini-pro": (0.125, 0.375),
    "google/gemini-pro-1.5": (1.25, 5.00),
    
    # Cohere models
    "cohere/command-r-plus": (3.00, 15.00),
    "cohere/command-r": (0.50, 1.50),
    
    # DeepSeek
    "deepseek/deepseek-chat": (0.14, 0.28),
    "deepseek/deepseek-coder": (0.14, 0.28),
}

# Aliases
MODEL_PRICING["gpt-4o"] = MODEL_PRICING["openai/gpt-4o"]
MODEL_PRICING["gpt-4o-mini"] = MODEL_PRICING["openai/gpt-4o-mini"]
MODEL_PRICING["claude-3.5-sonnet"] = MODEL_PRICING["anthropic/claude-3.5-sonnet"]
MODEL_PRICING["claude-3-haiku"] = MODEL_PRICING["anthropic/claude-3-haiku"]


@dataclass
class UsageRecord:
    """Record of a single LLM call"""
    timestamp: datetime
    model: str
    input_tokens: int
    output_tokens: int
    cost: float
    
    @property
    def total_tokens(self) -> int:
        return self.input_tokens + self.output_tokens


@dataclass
class CostTracker:
    """
    Tracks LLM usage and costs.
    
    Usage:
        tracker = CostTracker(limit=10.0)
        tracker.add(0.001, "gpt-4o-mini", 100, 50)
        print(f"Total: ${tracker.total_cost:.4f}")
        print(f"Remaining: ${tracker.remaining:.4f}")
    """
    
    limit: float = 10.0
    records: List[UsageRecord] = field(default_factory=list)
    
    @property
    def total_cost(self) -> float:
        """Total cost incurred"""
        return sum(r.cost for r in self.records)
    
    @property
    def remaining(self) -> float:
        """Remaining budget"""
        return max(0.0, self.limit - self.total_cost)
    
    @property
    def total_tokens(self) -> int:
        """Total tokens used"""
        return sum(r.total_tokens for r in self.records)
    
    @property
    def total_input_tokens(self) -> int:
        """Total input tokens used"""
        return sum(r.input_tokens for r in self.records)
    
    @property
    def total_output_tokens(self) -> int:
        """Total output tokens used"""
        return sum(r.output_tokens for r in self.records)
    
    def set_limit(self, limit: float):
        """Set the cost limit"""
        self.limit = limit
    
    def add(
        self,
        cost: float,
        model: str,
        input_tokens: int,
        output_tokens: int,
    ) -> UsageRecord:
        """
        Record a usage.
        
        Args:
            cost: Cost in USD
            model: Model name
            input_tokens: Number of input tokens
            output_tokens: Number of output tokens
            
        Returns:
            The created UsageRecord
        """
        record = UsageRecord(
            timestamp=datetime.now(),
            model=model,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            cost=cost,
        )
        self.records.append(record)
        return record
    
    def can_afford(self, estimated_cost: float) -> bool:
        """Check if an estimated cost can be afforded"""
        return self.remaining >= estimated_cost
    
    def estimate_cost(
        self,
        model: str,
        input_tokens: int,
        max_output_tokens: int = 1000,
    ) -> float:
        """
        Estimate cost for a request.
        
        Args:
            model: Model name
            input_tokens: Number of input tokens
            max_output_tokens: Maximum expected output tokens
            
        Returns:
            Estimated cost in USD
        """
        model_key = model.lower()
        pricing = MODEL_PRICING.get(model_key, (0.15, 0.6))
        input_price, output_price = pricing
        
        return (input_tokens * input_price / 1_000_000) + (max_output_tokens * output_price / 1_000_000)
    
    def get_summary(self) -> Dict:
        """Get usage summary"""
        model_costs: Dict[str, float] = {}
        model_tokens: Dict[str, int] = {}
        
        for record in self.records:
            model = record.model
            model_costs[model] = model_costs.get(model, 0) + record.cost
            model_tokens[model] = model_tokens.get(model, 0) + record.total_tokens
        
        return {
            "total_cost": self.total_cost,
            "remaining": self.remaining,
            "limit": self.limit,
            "total_tokens": self.total_tokens,
            "total_requests": len(self.records),
            "by_model": {
                model: {
                    "cost": model_costs[model],
                    "tokens": model_tokens[model],
                }
                for model in model_costs
            },
        }
    
    def reset(self):
        """Reset all records"""
        self.records.clear()


# Global cost tracker
_global_tracker: Optional[CostTracker] = None


def get_cost_tracker() -> CostTracker:
    """Get the global cost tracker"""
    global _global_tracker
    if _global_tracker is None:
        _global_tracker = CostTracker()
    return _global_tracker


def reset_cost_tracker():
    """Reset the global cost tracker"""
    global _global_tracker
    _global_tracker = CostTracker()

File: sdk/rust/src/cost.rs
//! Cost tracking for LLM usage

use chrono::{DateTime, Utc};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};

/// Model pricing in USD per million tokens (input, output)
pub struct ModelPricing;

impl ModelPricing {
    pub fn get(model: &str) -> (f64, f64) {
        match model.to_lowercase().as_str() {
            "openai/gpt-4o" => (2.5, 10.0),
            "openai/gpt-4o-mini" | "gpt-4o-mini" => (0.15, 0.6),
            "openai/gpt-4-turbo" => (10.0, 30.0),
            "openai/o1-preview" => (15.0, 60.0),
            "openai/o1-mini" => (3.0, 12.0),
            "anthropic/claude-3.5-sonnet" | "claude-3.5-sonnet" => (3.0, 15.0),
            "anthropic/claude-3-haiku" | "claude-3-haiku" => (0.25, 1.25),
            "anthropic/claude-3-opus" => (15.0, 75.0),
            "meta-llama/llama-3.1-70b-instruct" => (0.52, 0.75),
            "mistralai/mixtral-8x7b-instruct" => (0.24, 0.24),
            "google/gemini-pro" => (0.125, 0.375),
            "deepseek/deepseek-chat" => (0.14, 0.28),
            _ => (0.15, 0.6), // Default to gpt-4o-mini pricing
        }
    }

    pub fn calculate_cost(model: &str, input_tokens: u64, output_tokens: u64) -> f64 {
        let (input_price, output_price) = Self::get(model);
        (input_tokens as f64 * input_price / 1_000_000.0)
            + (output_tokens as f64 * output_price / 1_000_000.0)
    }
}

/// Record of a single LLM call
#[derive(Clone, Debug)]
pub struct UsageRecord {
    pub timestamp: DateTime<Utc>,
    pub model: String,
    pub input_tokens: u64,
    pub output_tokens: u64,
    pub cost: f64,
}

impl UsageRecord {
    pub fn total_tokens(&self) -> u64 {
        self.input_tokens + self.output_tokens
    }
}

/// Tracks LLM usage and costs
#[derive(Clone)]
pub struct CostTracker {
    inner: Arc<Mutex<CostTrackerInner>>,
}

struct CostTrackerInner {
    limit: f64,
    records: Vec<UsageRecord>,
}

impl Default for CostTracker {
    fn default() -> Self {
        Self::new(10.0)
    }
}

impl CostTracker {
    /// Create a new cost tracker with the given limit
    pub fn new(limit: f64) -> Self {
        Self {
            inner: Arc::new(Mutex::new(CostTrackerInner {
                limit,
                records: Vec::new(),
            })),
        }
    }

    /// Get the cost limit
    pub fn limit(&self) -> f64 {
        self.inner.lock().unwrap().limit
    }

    /// Set the cost limit
    pub fn set_limit(&self, limit: f64) {
        self.inner.lock().unwrap().limit = limit;
    }

    /// Get total cost incurred
    pub fn total_cost(&self) -> f64 {
        self.inner
            .lock()
            .unwrap()
            .records
            .iter()
            .map(|r| r.cost)
            .sum()
    }

    /// Get remaining budget
    pub fn remaining(&self) -> f64 {
        let inner = self.inner.lock().unwrap();
        let total: f64 = inner.records.iter().map(|r| r.cost).sum();
        (inner.limit - total).max(0.0)
    }

    /// Get total tokens used
    pub fn total_tokens(&self) -> u64 {
        self.inner
            .lock()
            .unwrap()
            .records
            .iter()
            .map(|r| r.total_tokens())
            .sum()
    }

    /// Record a usage
    pub fn add(&self, model: &str, input_tokens: u64, output_tokens: u64, cost: f64) -> UsageRecord {
        let record = UsageRecord {
            timestamp: Utc::now(),
            model: model.to_string(),
            input_tokens,
            output_tokens,
            cost,
        };

        self.inner.lock().unwrap().records.push(record.clone());
        record
    }

    /// Check if an estimated cost can be afforded
    pub fn can_afford(&self, estimated_cost: f64) -> bool {
        self.remaining() >= estimated_cost
    }

    /// Estimate cost for a request
    pub fn estimate_cost(&self, model: &str, input_tokens: u64, max_output_tokens: u64) -> f64 {
        ModelPricing::calculate_cost(model, input_tokens, max_output_tokens)
    }

    /// Get usage summary by model
    pub fn summary(&self) -> HashMap<String, (f64, u64)> {
        let inner = self.inner.lock().unwrap();
        let mut summary: HashMap<String, (f64, u64)> = HashMap::new();

        for record in &inner.records {
            let entry = summary.entry(record.model.clone()).or_insert((0.0, 0));
            entry.0 += record.cost;
            entry.1 += record.total_tokens();
        }

        summary
    }

    /// Reset all records
    pub fn reset(&self) {
        self.inner.lock().unwrap().records.clear();
    }
}
